lib/tree.rb

```diff
@@ -16,40 +16,137 @@ def initialize
+  # Space Complexity: o(n)
   def add(key, value)
-    raise NotImplementedError
+    new_node = TreeNode.new(key, value)
+    
+    if @root.nil?
+      @root = new_node
+      return
+    end
+
+    return add_helper(@root, new_node)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper(current, new_node)
+    return new_node if current.nil?
+
+    if new_node.key <= current.key
+      current.left = add_helper(current.left, new_node)
+    else
+      current.right = add_helper(current.right, new_node)
+    end
+    return current
+  end
+
+  # Time Complexity: o(log n) if it's balanced; worst case is o(n)
+  # Space Complexity: o(n)
   def find(key)
-    raise NotImplementedError
+    return nil if @root.nil?
+    return find_helper(@root, key)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def find_helper(current, key)
+    return current.value if current.key == key
+
+    if key <= current.key
+      return nil if current.left == nil
+      current.left = find_helper(current.left, key)
+    else
+      return nil if current.right == nil
+      current.right = find_helper(current.right, key)
+    end
+
+  end
+
+  # Time Complexity: o(n)
+  # Space Complexity: o(n)
   def inorder
-    raise NotImplementedError
+    return [] if @root.nil?
+    result = []
+    return inorder_helper(@root, result)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(node, result)
+    return if node.nil?
+  
+    if node.left
+      inorder_helper(node.left, result) 
+    end
+    result << { :key => node.key, :value => node.value }
+    if node.right
+      inorder_helper(node.right, result) 
+    end
+
+    return result
+  end
+
+  # Time Complexity: o(n)
+  # Space Complexity: o(n)
   def preorder
-    raise NotImplementedError
+    return [] if @root.nil?
+    result = []
+    return preorder_helper(@root, result)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(node, result)
+    return if node.nil?
+  
+    result << { :key => node.key, :value => node.value }
+
+    if node.left
+      preorder_helper(node.left, result) 
+    end
+    # result << { :key => node.key, :value => node.value }
+    if node.right
+      preorder_helper(node.right, result) 
+    end
+
+    return result
+  end
+
+  # Time Complexity: o(n)
+  # Space Complexity: o(n)
   def postorder
-    raise NotImplementedError
+    return [] if @root.nil?
+    result = []
+    return postorder_helper(@root, result)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def postorder_helper(node, result)
+    return if node.nil?
+  
+    if node.left
+      postorder_helper(node.left, result) 
+    end
+    # result << { :key => node.key, :value => node.value }
+    if node.right
+      postorder_helper(node.right, result) 
+    end
+    result << { :key => node.key, :value => node.value }
+
+    return result
+  end
+
+  # Time Complexity: o(log n)
+  # Space Complexity: o(n)
   def height
```

Comment:
👍 The time complexity is O(n) since you visit each node, and space complexity is O(log n) if the tree is balanced and O(n) if it's unbalanced.